File: src/main/scala/com/vc/easy/L671.scala
package com.vc.easy


class TreeNode671(var _value: Int) {
  var value: Int = _value
  var left: TreeNode671 = null
  var right: TreeNode671 = null
}

object L671 {
  def findSecondMinimumValue(root: TreeNode671): Int = {
    var min = Int.MaxValue
    var res = Int.MaxValue
    def solve(root: TreeNode671): Unit = {
      if(root != null) {
        if(root.value < min) {
          res = min
          min = root.value
        }
        else if(root.value < res && root.value != min) {
          res = root.value
        }
        solve(root.left)
        solve(root.right)
      }
    }
    solve(root)
    if(res == Int.MaxValue) -1 else res
  }
}

File: src/main/java/com/vc/easy/L671.java
package com.vc.easy;

import com.vc.hard.TreeNode;

class L671 {
    private long min1 = Long.MAX_VALUE, min2 = Long.MAX_VALUE;
    public int findSecondMinimumValue(TreeNode root) {
        helper(root);
        return min2 == Long.MAX_VALUE ? -1 : (int)min2;
    }

    private void helper(TreeNode root) {
        if(root == null) return;

        if(min1 > root.val) min1 = root.val;
        else if(min2 > root.val && min1 != root.val) min2 = root.val;

        helper(root.left);
        helper(root.right);
    }
}
